File: renew.go
package main

import (
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
)

// RenewalResult holds the outcome for a single loan after the POST.
type RenewalResult struct {
    Loan       Loan
    Success    bool
    NewDueDate time.Time
    ErrMsg     string
}

// renewAll sends a "Renew All" POST to Finna using the shared client.
// If DRY_RUN=true, it simulates success without sending any POST.
func renewAll(cfg Config, csrf string, loans []Loan, client *http.Client) ([]RenewalResult, error) {

    // ───────────────────────────────────────────────────────────────
    // DRY RUN MODE — simulate success without touching Finna
    // ───────────────────────────────────────────────────────────────
    if cfg.DryRun {
        results := make([]RenewalResult, len(loans))
        for i, l := range loans {
            results[i] = RenewalResult{
                Loan:       l,
                Success:    true,
                NewDueDate: l.DueDate, // unchanged in dry-run
            }
        }
        return results, nil
    }

    // ───────────────────────────────────────────────────────────────
    // Build POST body
    // ───────────────────────────────────────────────────────────────
    body := buildRenewalForm(csrf, loans)

    req, err := http.NewRequest("POST", checkedOutURL, strings.NewReader(body))
    if err != nil {
        return nil, err
    }
    setBaseHeaders(req)
    req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
    req.Header.Set("Referer", checkedOutURL)

    // Renewal can be slow when many items exist.
    client.Timeout = 60 * time.Second

    // ───────────────────────────────────────────────────────────────
    // Send POST with retry logic
    // ───────────────────────────────────────────────────────────────
    resp, err := doRequestWithRetry(client, req)
    if err != nil {
        return nil, fmt.Errorf("POST renewal: %w", err)
    }
    defer resp.Body.Close()

    // Session expired → caller will re-login and retry
    if isLoginPage(resp.Request.URL) {
        return nil, ErrSessionExpired
    }

    if resp.StatusCode != 200 {
        return nil, fmt.Errorf("renewal POST returned HTTP %d", resp.StatusCode)
    }

    // ───────────────────────────────────────────────────────────────
    // Parse renewal results
    // ───────────────────────────────────────────────────────────────
    return parseRenewalResponse(resp.Body, loans)
}

// buildRenewalForm constructs the URL-encoded POST body.
func buildRenewalForm(csrf string, loans []Loan) string {
    v := url.Values{}
    v.Set("csrf", csrf)
    v.Set("renewAll", "1")
    v.Set("selectAll", "on")

    for _, loan := range loans {
        v.Add("renewSelectedIDS[]", loan.ID)
        v.Add("selectAllIDS[]", loan.ID)
        v.Add("renewAllIDS[]", loan.ID)
    }

    return v.Encode()
}

// parseRenewalResponse reads the HTML from the renewal POST response and
// matches each loan row to a success or failure outcome.
func parseRenewalResponse(body io.Reader, loans []Loan) ([]RenewalResult, error) {
    doc, err := goquery.NewDocumentFromReader(body)
    if err != nil {
        return nil, fmt.Errorf("parse renewal HTML: %w", err)
    }

    // Map loan ID → original Loan
    loanByID := make(map[string]Loan, len(loans))
    for _, l := range loans {
        loanByID[l.ID] = l
    }

    var results []RenewalResult

    doc.Find("tr.myresearch-row").Each(func(_ int, row *goquery.Selection) {
        id, ok := row.Find(`input[name="renewSelectedIDS[]"]`).First().Attr("value")
        if !ok || id == "" {
            return
        }

        loan := loanByID[id]
        result := RenewalResult{Loan: loan}

        statusCol := row.Find(".status-column")
        statusText := statusCol.Text()

        // ───────────────────────────────────────────────────────────────
        // SUCCESS
        // ───────────────────────────────────────────────────────────────
        if statusCol.Find(".alert-success").Length() > 0 {
            result.Success = true

            // Try to parse new due date
            newDue := parseDueDate(statusText)

            // Cloudflare sometimes strips HTML → fallback to original due date
            if newDue.IsZero() {
                newDue = loan.DueDate
            }

            result.NewDueDate = newDue
            results = append(results, result)
            return
        }

        // ───────────────────────────────────────────────────────────────
        // FAILURE
        // ───────────────────────────────────────────────────────────────
        if statusCol.Find(".alert-danger").Length() > 0 {
            result.Success = false
            result.ErrMsg = strings.TrimSpace(statusCol.Find(".alert-danger").Text())
            results = append(results, result)
            return
        }

        // ───────────────────────────────────────────────────────────────
        // UNKNOWN (should not happen)
        // ───────────────────────────────────────────────────────────────
        result.Success = false
        result.ErrMsg = "no renewal status in response"
        results = append(results, result)
    })

    return results, nil
}

File: src/server/finna/renew.ts
import * as cheerio from "cheerio";
import type { Loan, RenewalResult, FinnaSession } from "./types";
import { fetchWithRetry } from "./client";
import { parseDueDate } from "./loans";

export async function renewSelected(
  session: FinnaSession,
  csrf: string,
  loanIds: string[],
  allLoans: Loan[],
  renewalUrl: string,
): Promise<RenewalResult[]> {
  const selected = allLoans.filter((l) => loanIds.includes(l.id));
  if (selected.length === 0) return [];

  const params = new URLSearchParams();
  params.set("csrf", csrf);
  for (const loan of selected) {
    params.append("renewSelectedIDS[]", loan.id);
  }
  // Include all IDs in selectAllIDS/renewAllIDS (Finna expects them)
  for (const loan of allLoans) {
    params.append("selectAllIDS[]", loan.id);
    params.append("renewAllIDS[]", loan.id);
  }

  return postRenewal(session, params.toString(), selected, renewalUrl);
}

export async function renewAll(
  session: FinnaSession,
  csrf: string,
  loans: Loan[],
  renewalUrl: string,
): Promise<RenewalResult[]> {
  const body = buildRenewalForm(csrf, loans);
  return postRenewal(session, body, loans, renewalUrl);
}

async function postRenewal(
  session: FinnaSession,
  body: string,
  loans: Loan[],
  renewalUrl: string,
): Promise<RenewalResult[]> {
  const resp = await fetchWithRetry(renewalUrl, {
    method: "POST",
    cookies: session.cookies,
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      Referer: renewalUrl,
    },
    body,
  });

  if (
    resp.url.includes("/MyResearch/Login") ||
    resp.url.includes("/MyResearch/UserLogin")
  ) {
    throw new Error("SESSION_EXPIRED");
  }

  if (resp.status !== 200) {
    throw new Error(`Renewal POST returned HTTP ${resp.status}`);
  }

  const html = await resp.text();
  return parseRenewalResponse(html, loans);
}

function buildRenewalForm(csrf: string, loans: Loan[]): string {
  const params = new URLSearchParams();
  params.set("csrf", csrf);
  params.set("renewAll", "1");
  params.set("selectAll", "on");

  for (const loan of loans) {
    params.append("renewSelectedIDS[]", loan.id);
    params.append("selectAllIDS[]", loan.id);
    params.append("renewAllIDS[]", loan.id);
  }

  return params.toString();
}

function parseRenewalResponse(html: string, loans: Loan[]): RenewalResult[] {
  const $ = cheerio.load(html);

  const loanById = new Map(loans.map((l) => [l.id, l]));
  const results: RenewalResult[] = [];

  $("tr.myresearch-row").each((_, row) => {
    const $row = $(row);
    const id = $row
      .find('input[name="renewSelectedIDS[]"]')
      .first()
      .attr("value");
    if (!id) return;

    const loan = loanById.get(id);
    if (!loan) return;

    const $statusCol = $row.find(".status-column");
    const statusText = $statusCol.text();

    if ($statusCol.find(".alert-success").length > 0) {
      let newDue = parseDueDate(statusText);
      if (newDue.getTime() === 0) newDue = loan.dueDate;

      results.push({ loan, success: true, newDueDate: newDue });
      return;
    }

    if ($statusCol.find(".alert-danger").length > 0) {
      const errMsg = $statusCol.find(".alert-danger").text().trim();
      results.push({ loan, success: false, errorMessage: errMsg });
      return;
    }

    results.push({
      loan,
      success: false,
      errorMessage: "No renewal status in response",
    });
  });

  return results;
}
